# Nama      : Brian Mohamad Safiudin
# NIM       : 2141720133
# Kelas     : TI - 3F

---
# **Tugas Praktikum Perceptron dan ANN - Pertemuan 9**
---
##### Lakukan klasifikasi pada data MNIST dengan menggunakan model ANN
##### Anda diperbolehkan melakukan eksplorasi terhadap,
- Metode pra pengolahan
- Pemilihan fitur
- Arsitektur ANN
- Fungsi Aktiviasi
##### ANN diimplementasikan dengan menggunakan tensorflow.
##### DIKERJAKAN SECARA BERKELOMPOK
##### JELASKAN HASIL YANG ANDA DAPATKAN,
- AKURASI
- CONFUSION MATRIX
- KONFIGURASI MODEL --> MULAI DARI PRA PENGOLAHAN SAMPAI ARSITEKTUR ANN

---
# **Import Library**
---

In [1]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
import matplotlib.pyplot as plt

---
# **Muat data MNIST**
---

In [2]:
mnist = keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

11490434/11490434 [==============================] - 54s 5us/step


---
# **Pra Pengolahan Data**
---

In [3]:
train_images = train_images.reshape(-1, 28*28)
test_images = test_images.reshape(-1, 28*28)

In [4]:
train_images = train_images / 255.0
test_images = test_images / 255.0

---
# **Arsitektur ANN**
---

In [5]:
model = keras.Sequential([
    layers.Dense(128, activation='relu', input_shape=(784,)),
    layers.Dense(64, activation='relu'),
    layers.Dense(10, activation='softmax')
])

---
# **Kompilasi Model**
---

In [6]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

---
# **Pelatihan Model**
---

In [7]:
model.fit(train_images, train_labels, epochs=10, validation_split=0.2)

Epoch 1/10
1500/1500 [==============================] - 7s 4ms/step - loss: 0.2682 - accuracy: 0.9225 - val_loss: 0.1434 - val_accuracy: 0.9586
Epoch 2/10
1500/1500 [==============================] - 6s 4ms/step - loss: 0.1137 - accuracy: 0.9660 - val_loss: 0.1158 - val_accuracy: 0.9651
Epoch 3/10
1500/1500 [==============================] - 6s 4ms/step - loss: 0.0800 - accuracy: 0.9754 - val_loss: 0.0975 - val_accuracy: 0.9696
Epoch 4/10
1500/1500 [==============================] - 6s 4ms/step - loss: 0.0588 - accuracy: 0.9817 - val_loss: 0.0953 - val_accuracy: 0.9699
Epoch 5/10
1500/1500 [==============================] - 6s 4ms/step - loss: 0.0459 - accuracy: 0.9856 - val_loss: 0.1036 - val_accuracy: 0.9703
Epoch 6/10
1500/1500 [==============================] - 6s 4ms/step - loss: 0.0377 - accuracy: 0.9876 - val_loss: 0.0897 - val_accuracy: 0.9754
Epoch 7/10
1500/1500 [==============================] - 6s 4ms/step - loss: 0.0309 - accuracy: 0.9893 - val_loss: 0.1004 - val_accuracy:

---
# **Evaluasi Model**
---

In [8]:
test_loss, test_accuracy = model.evaluate(test_images, test_labels)
print("Test accuracy: {:.4f}".format(test_accuracy))

313/313 [==============================] - 1s 2ms/step - loss: 0.0925 - accuracy: 0.9756
Test accuracy: 0.9756


---
# **Confusion Matrix**
---

In [9]:
from sklearn.metrics import confusion_matrix

predicted_labels = model.predict(test_images)
predicted_labels = [tf.argmax(label).numpy() for label in predicted_labels]

confusion = confusion_matrix(test_labels, predicted_labels)
print(confusion)

313/313 [==============================] - 1s 2ms/step
[[ 971    0    4    1    0    1    1    0    1    1]
 [   0 1130    2    1    0    0    2    0    0    0]
 [   2    2 1016    2    3    0    1    3    3    0]
 [   1    0    5  992    0    3    0    5    4    0]
 [   2    0    4    0  954    1    6    5    1    9]
 [   3    0    0   11    0  870    4    1    2    1]
 [   5    2    0    1    3    3  943    0    1    0]
 [   2    5    9    5    0    0    1  995    4    7]
 [   3    1   10   13    2    5    4    4  930    2]
 [   2    5    1   15   13    7    1    6    4  955]]


---
# **Hasil Yang Saya Dapatkan**
---
- Confusion Matrix
```
313/313 [==============================] - 1s 2ms/step - loss: 0.0925 - accuracy: 0.9756
```
- Akurasi model pada data pengujian adalah sekitar 97.56%.
```
313/313 [==============================] - 1s 2ms/step
[[ 971    0    4    1    0    1    1    0    1    1]
 [   0 1130    2    1    0    0    2    0    0    0]
 [   2    2 1016    2    3    0    1    3    3    0]
 [   1    0    5  992    0    3    0    5    4    0]
 [   2    0    4    0  954    1    6    5    1    9]
 [   3    0    0   11    0  870    4    1    2    1]
 [   5    2    0    1    3    3  943    0    1    0]
 [   2    5    9    5    0    0    1  995    4    7]
 [   3    1   10   13    2    5    4    4  930    2]
 [   2    5    1   15   13    7    1    6    4  955]]
```
##### Konfigurasi model
- Pra Pengolahan Data: Citra-citra dari dataset MNIST dilatih setelah diubah menjadi vektor 1D dan dinormalisasi agar berada dalam rentang 0 hingga 1.  

Arsitektur ANN:
  - Hidden layer pertama memiliki 128 unit dengan aktivasi ReLU.
  - Hidden layer kedua memiliki 64 unit dengan aktivasi ReLU.
  - Output layer memiliki 10 unit (sesuai dengan jumlah kelas digit) dengan aktivasi softmax.  
  - Model menggunakan arsitektur ANN dengan tiga lapisan, yaitu satu lapisan input, dua lapisan tersembunyi, dan satu lapisan output.
  - Input layer memiliki 784 unit (28x28 piksel) dan fungsi aktivasi ReLU.

##### Pelatihan Model: 
- Hasil: Model mencapai akurasi sekitar 97.56% pada data pengujian, yang menunjukkan kemampuan yang sangat baik dalam mengklasifikasikan digit-digit MNIST.
- Model dilatih selama 10 epoch dengan menggunakan optimizer 'adam', loss function 'sparse_categorical_crossentropy' (karena ini adalah masalah klasifikasi multikelas), dan metrik akurasi.  